wp-cli/vip-go-convert-to-utf8mb4.php

```diff
@@ -0,0 +1,348 @@
+	 *
+	 * @param string $table
+	 * @return array
+	 */
+	private function convert_masquerading_columns( $table ) {
+		global $wpdb;
+
+		$columns = $wpdb->get_results( "SHOW COLUMNS FROM $table;" );
+
+		$cols_converted = array();
+
+		foreach ( $columns as $col ) {
+			$cols_converted[ $col->Field ] = $this->maybe_convert_column( $col, $table );
+		}
+
+		return $cols_converted;
+	}
+
+	/**
+	 * Convert latin1 columns that actually contain utf8
+	 *
+	 * @return mixed
+	 */
+	private function maybe_convert_column( $col, $table ) {
+		global $wpdb;
+		$from_type = $to_type = null;
+
+		foreach ( self::TYPE_MAPPING as $from => $to ) {
+			// Most will be exact matches, except for varchar and enum
+			if (
+				$col->Type === $from ||
+				( 'enum' === $from && 0 === stripos( $col->Type, $from ) ) ||
+				( 'varchar' === $from && 0 === stripos( $col->Type, $from ) )
+			) {
+				$from_type = $from;
+				$to_type = $to;
+				break;
+			}
+		}
+
+		unset( $from, $to );
+
+		// Carry on, we don't care about this column
+		if ( is_null( $from_type ) || is_null( $to_type ) ) {
+			return false;
+		}
+
+		// enums are special
+		if ( 0 === stripos( $from_type, 'enum' ) ) {
+			// If we can't find the options, something is very wrong
+			if ( ! preg_match( '#' . preg_quote( $from_type ) . '\(([^\)]+)\)#i', $col->Type, $options ) ) {
+				return false;
+			}
+
+			$options       = array_pop( $options );
+			$null_not_null = 'yes' === strtolower( $col->Null ) ? 'NULL' : 'NOT NULL';
+
+			WP_CLI::line( "Converting column {$col->Field}" );
+
+			$convert = $wpdb->query( $wpdb->prepare( 'ALTER TABLE %1$s CHANGE %2$s %2$s ENUM(' . $options . ') CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci %3$s DEFAULT "%4$s"', $table, $col->Field, $null_not_null, $col->Default ) );
+
+			WP_CLI::line( "Finished converting column {$col->Field}" );
+
+			return array(
+				'convert' => $convert,
+				'restore' => $convert,
+			);
+		}
+
+		// Maintain varchar length
+		if ( 0 === stripos( $from_type, 'varchar' ) ) {
+			// If we can't find the length, something is very wrong
+			if ( ! preg_match( '#' . preg_quote( $from_type ) . '\(([\d]+)\)#i', $col->Type, $length ) ) {
+				return false;
+			}
+
+			$length = array_pop( $length );
+
+			if ( ! is_numeric( $length ) ) {
+				return false;
+			}
+
+			$from_type .= "({$length})";
+			$to_type   .= "({$length})";
+		}
+
+		$null_not_null = 'yes' === strtolower( $col->Null ) ? ' NULL' : ' NOT NULL';
+
+		// On with it!
+		WP_CLI::line( "Converting column {$col->Field}" );
+
+		$pattern = 'ALTER TABLE %1$s CHANGE %2$s %2$s %3$s %4$s %5$s DEFAULT "%6$s"';
+		$convert = $wpdb->query( $wpdb->prepare( $pattern, $table, $col->Field, $to_type, '', $null_not_null, $col->Default ) );
+		$restore = $wpdb->query( $wpdb->prepare( $pattern, $table, $col->Field, $from_type, 'CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci', $null_not_null, $col->Default ) );
```

Comment:
Minor: Not totally clear why we need to convert and then restore; inline comments would be helpful.